### PR TITLE
Add a KVM feature and use it by default

### PR DESCRIPTION
**Pull request by liuw**
I wrote some code to demonstrate the idea put forward in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1421.

### CODE REVIEW

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -1446,8 +1446,7 @@ pub fn test_vm() {
     let load_addr = GuestAddress(0x1000);
     let mem = GuestMemoryMmap::from_ranges(&[(load_addr, mem_size)]).unwrap();
 
-    let kvm = hypervisor::kvm::KvmHypervisor::new().unwrap();
-    let hv: Arc<dyn hypervisor::HypervisorTrait> = Arc::new(kvm);
+    let hv: Arc<dyn hypervisor::HypervisorTrait> = Arc::new(hypervisor::Hypervisor::new().unwrap());
```
Comment (sboeuf) [inline vmm/src/vm.rs:1449]: No I wouldn't proceed this way. Instead, we should have two different hypervisor initializations gated by a feature. Both would return a structure implementing the Hypervisor trait.

File: hypervisor/src/lib.rs [hunk 2]
```diff
@@ -25,6 +25,8 @@ extern crate thiserror;
 #[macro_use]
 extern crate anyhow;
 
+pub use kvm::KvmHypervisor as Hypervisor;
```
Comment (sboeuf) [inline hypervisor/src/lib.rs:29]: I don't see the reason why we would do that. It's important that we keep Hypervisor as the trait, and then HypervHypervisor and KvmHypervisor will implement it. The caller (vmm crate) will initialize the right structure depending on the rust feature present at build time. Once initialized, the structure will be used as the Hypervisor trait for the generic code shared by both features (kvm and hyperv).

File: Cargo.toml [hunk 3]
```diff
@@ -35,12 +35,13 @@ serde_json = "1.0.56"
 net_util = { path = "net_util" }
 
 [features]
-default = ["acpi", "pci", "cmos"]
+default = ["acpi", "pci", "cmos", "kvm"]
```
Comment (sboeuf) [inline Cargo.toml:38]: looks good!

File: arch/src/aarch64/fdt.rs [hunk 4]
```diff
@@ -587,7 +587,7 @@ mod tests {
         .collect();
 
         let kvm = hypervisor::kvm::KvmHypervisor::new().unwrap();
-        let hv: Arc<dyn hypervisor::Hypervisor> = Arc::new(kvm);
+        let hv: Arc<dyn hypervisor::HypervisorTrait> = Arc::new(kvm);
```
Comment (sboeuf) [inline arch/src/aarch64/fdt.rs]: As explained in other comments, I don't think we need to re-export `KvmHypervisor` as `Hypervisor`, therefore we shouldn't need to rename the `Hypervisor` trait into `HypervisorTrait` trait.
Also, we should never use `Trait` in a trait's name, as it is redundant.

File: hypervisor/src/lib.rs (hunk 2 above)
Comment (liuw) [reply hypervisor/src/lib.rs:29]: But that means code in VMM crate will need to have 

```
#[cfg(feature = "kvm")]
let hv = hypervisor::kvm::KvmHypervisor::new().unwrap();
#[cfg(feature = "hyperv")]
let hv = hypervisor::hyperv::HyperVHypervisor::new().unwrap();
let hv: Arc<dyn hypervisor::Hypervisor> = Arc::new(hv);
```

And this snippet needs to be duplicated in _all_ the present and future test cases. That looks rather cumbersome, isn't it?

File: vmm/src/vm.rs (hunk 1 above)
Comment (liuw) [reply vmm/src/vm.rs:1449]: To be clear: do you want these two different hypervisor initializations be put in the `vmm` crate or the `hypervisor` crate?

File: arch/src/aarch64/fdt.rs (hunk 4 above)
Comment (liuw) [reply arch/src/aarch64/fdt.rs]: I agree. It is just that Rust (rightfully) wouldn't allow me to export a trait and a concrete type with the same name. This issue may become moot once we settle other issues.

File: vmm/src/vm.rs (hunk 1 above)
Comment (sboeuf) [reply vmm/src/vm.rs:1449]: The hypervisor crate should be agnostic from initialization and how things are being used. That's why I would suggest the two different hypervisor initialization would be put in the `vmm` crate. They would be gated by feature.

File: hypervisor/src/lib.rs (hunk 2 above)
Comment (sboeuf) [reply hypervisor/src/lib.rs:29]: No that's fine. From the vmm crate we want to be able to identify what's kvm vs hyperv specific. We don't want this kind of thing to be pushed down to the hypervisor crate.

Comment (liuw) [reply hypervisor/src/lib.rs:29]: Sure. This is also fine by me.

**Comment by liuw** [comment]
> 
> 
> I think you're missing the feature check from the code in `vmm/src/vm.rs`.

Do you mean adding `#[cfg(feature = "kvm")]` before `let kvm = ...`? I plan to add all the feature checks in a later patch. There are more than one place that needs fixing. For now everything should still work because kvm is enabled by default.

**Comment by russell-islam** [comment]
Looks good to me!

**Comment by liuw** [comment]
```
error: Failed to delete invalidated or incompatible incremental compilation session directory contents `/cloud-hypervisor/target/debug/incremental/vhost_user_net-22dwtdnrq4v3n/s-fp34jikm4e-m7oy0t-working/dep-graph.bin`: No such file or directory (os error 2).
```

The error doesn't seem to be related to my changes.

**Pull request merged**